pykafka/producer.py

```diff
@@ -241,3 +237,326 @@ def produce(self, messages):
+            new messages (block) or throw an error. If True, this setting
+            indicates we should block until space is available in the queue.
+        :type block_on_queue_full: bool
+        """
+        self._cluster = cluster
+        self._topic = topic
+        self._partitioner = partitioner
+        self._compression = compression
+        self._max_retries = max_retries
+        self._retry_backoff_ms = retry_backoff_ms
+        self._required_acks = required_acks
+        self._ack_timeout_ms = ack_timeout_ms
+        self._max_queued_messages = max_queued_messages
+        self._linger_ms = linger_ms
+        self._block_on_queue_full = block_on_queue_full
+
+        self._running = False
+        self.start()
+
+    def start(self):
+        """Connect to brokers and start worker threads"""
+        if not self._running:
+            self._setup_internal_producer()
+            self._setup_workers()
+            self._running = True
+
+    def stop(self):
+        """Mark as stopped and wait for all inflight messages to send"""
+        self._running = False
+        last_log = time.time()
+        while any(q_info.messages_inflight > 0
+                  for leader, q_info in self._workers_by_leader.iteritems()):
+            if time.time() - last_log >= .5:
+                log.info("Waiting for all worker threads to exit")
+                last_log = time.time()
+
+    def __enter__(self):
+        """Context manager entry point - start the producer"""
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Context manager exit point - stop the producer"""
+        self.stop()
+
+    def _setup_internal_producer(self):
+        """Instantiate the internal producer"""
+        self._producer = Producer(
+            self._cluster,
+            self._topic,
+            self._partitioner,
+            compression=self._compression,
+            max_retries=self._max_retries,
+            retry_backoff_ms=self._retry_backoff_ms,
+            required_acks=self._required_acks,
+            ack_timeout_ms=self._ack_timeout_ms,
+        )
+
+    def _setup_workers(self):
+        """Spawn workers for connected brokers
+
+        Creates one :class:`queue.Queue` instance and one worker thread per
+        broker. The queue holds messages waiting to be sent, and the worker
+        thread reads the end of the queue and sends requests.
+        """
+        # _workers_by_leader maps each broker id to a QueueInfo instance
+        # containing references to that broker's queue and its associated
+        # synchronization primitives
+        self._workers_by_leader = {}
+        for partition in self._topic.partitions.values():
+            if partition.leader.id not in self._workers_by_leader:
+                q_info = self._setup_queue(partition.leader)
+                self._workers_by_leader[partition.leader.id] = q_info
+
+    def _setup_queue(self, broker):
```

Comment:
The implementation no longer requires thread-per-broker because we already have a network thread for each broker. Adding more here just adds unnecessary complexity.